include/interrupts.h

```diff
@@ -39,6 +39,8 @@ void intr_event_remove_handler(intr_handler_t *ih);
 void intr_event_execute_handlers(intr_event_t *ie);
 void run_event_handlers(unsigned irq);
 
+bool during_intr_handler();
+
 #define intr_disable() __extension__ ({ asm("di"); })
```

Comment:
I'm aware that it's not part of this patch, but since it is in the context I'm going to comment anyway.
Interrupt disabling and enabling functions should also clobber memory so that the compiler is not tempted to reorder memory accesses.